### PR TITLE
Added capability for unique identifiers when deploying multiple space-scoped instances of the broker

### PR DESCRIPTION
**Pull request by mszcool**
When adding a service broker to a CF environment, Service IDs, Service Names and Plan IDs need to be unique for the entire Cloud Foundry environment.

In multi-tenant CF environment such as SAP Cloud Platform where multiple individual customers get their own organizations and spaces and don't have the permission to change global configurations, Service Brokers can only installed in a space-scoped fashion. In such cases, to enable customers making use of the Azure Meta Service Broker, the broker would need to automatically generate Service IDs, Service Names and Plan IDs which are unique to the entire CF environment to avoid conflicts with other customers who have an instance of the broker installed in their spaces with the option "--space-scoped", already.

Therefore, with this pull request, I've added a new option called SPACE_SCOPING_ENABLED. If Set to true, it will generate new Service IDs, Service Names and Plan IDs which are unique to the entire CF environment. Also display names are adjusted accordingly so that developers will be able to see, which services are installed space-scoped.

### CODE REVIEW

**Comment by zhongyi-zhang** [comment]
@mszcool thanks for this awesome contribution! I am on a vacation today. Will get it reviewed next Monday.

File: docs/how-admin-deploy-the-broker.md [hunk 1]
```diff
@@ -152,6 +153,10 @@
...
 
+      * `SPACE_SCOPING_ENABLED`
+
+          Needs to be set to `true` if the service broker will be deployed in multiple spaces of the same Cloud Foundry environment with the option `--space-scoped` specified when executing `cf create-service-broker` to register the service broker with the Cloud Controller. This option will ensure, that Service Names, Service IDs and Plan IDs are unique for the entire Cloud Foundry environment in case multiple instances of the broker are deployed in different spaces of the same CF environment. The reason for this is, that Service Names, Service IDs and Plan IDs need to be unique even if the broker is deployed with the option `--space-scoped` specified when executing `cf create-service-broker`. The default assumes, that one instance of the service broker is deployed for the entire Cloud Foundry environment, hence you should set this to `false` in most of the cases.
+
```
[the remark this answers is not in the document]
Comment (mszcool) [reply docs/how-admin-deploy-the-broker.md]: Hi, I'll add a note for avoiding the conflicting route with a sample-bash script that generates a proposal for a unique route-name. Enabling service access should not be necessary for --space-scoped brokers, at least that's the case in SAP CP. But, I've added a note to it.

File: lib/common/index.js [hunk 2]
```diff
@@ -242,6 +261,30 @@ module.exports.API_VERSION = {
...
+  
+  // Modifying the service catalog to ensure serivce names and IDs are unique for the space
+  if (scopingEnabled) {
+    // Reading the CF VCAP Application properties with the space ID
+    var vcapApplication = JSON.parse(process.env['VCAP_APPLICATION']);
+    var spaceId = vcapApplication.space_id;
+    var spaceName = vcapApplication.space_name;
+
+    // Update service names, service ids and plan ids to remain unique if broker is deployed in multiple spaces
+    serviceConfig.id = uuid.v4();
```
[the remark this answers is not in the document]
Comment (mszcool) [reply lib/common/index.js]: Working on the fix. Good catch, I totally missed out on this one, thanks!!!

[the remark this answers is not in the document]
Comment (mszcool) [reply lib/common/index.js]: I think what needs to happen is, using uuid.v5(serviceName, spaceId) for this one (so using the name of the service and the ID of the space as seed) while for the others it needs to be different.

File: lib/common/index.js [hunk 3]
```diff
@@ -242,6 +261,30 @@ module.exports.API_VERSION = {
...
+    serviceConfig.name = serviceConfig.name + '-' + spaceId; //brokerConfigurations.serviceBroker.spaceScopingNamePostfix.toLowerCase();
+    serviceConfig.description = serviceConfig.description + ' (' + spaceName + ')'; //brokerConfigurations.serviceBroker.spaceScopingNamePostfix + ')';
+    serviceConfig.plans.forEach(function (plan) {
+      plan.id = uuid.v4();
```
[the remark this answers is not in the document]
Comment (mszcool) [reply lib/common/index.js]: makes sense, will adopt approach accordingly.

File: manifest.yml [hunk 4]
```diff
@@ -11,13 +11,15 @@ applications:
...
     SECURITY_USER_PASSWORD: REPLACE-ME
+    SPACE_SCOPING_ENABLED: true | false
     AZURE_BROKER_DATABASE_PROVIDER: REPLACE-ME
     AZURE_BROKER_DATABASE_SERVER: REPLACE-ME
     AZURE_BROKER_DATABASE_USER: REPLACE-ME
     AZURE_BROKER_DATABASE_PASSWORD: REPLACE-ME
     AZURE_BROKER_DATABASE_NAME: REPLACE-ME
     AZURE_BROKER_DATABASE_ENCRYPTION_KEY: REPLACE-ME
 
+
```
[the remark this answers is not in the document]
Comment (mszcool) [reply manifest.yml]: happend accidentally

File: lib/common/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhongyi-zhang) [reply lib/common/index.js]: Yes, uuidV5 should work well in this case!

File: test/unit/common/utils-tests.js [hunk 5]
```diff
@@ -286,99 +287,248 @@ describe('Util', function() {
...
+          }],
+          'displayName': 'Standard'
+        }
+      }]
+    };
+
+    function copyTestServiceConfig() {
+      // Needs to copy the test service configuration to ensure its not overwritten as part of testing
+      // I know its not the most effective way, but it was the fastest way to get the tests completed, quickly
+      return JSON.parse(JSON.stringify(testServiceConfig));
```
[the remark this answers is not in the document]
Comment (zhongyi-zhang) [reply test/unit/common/utils-tests.js]: Well done!

**Comment by zhongyi-zhang** [comment]
@mszcool merged. Thanks again for the contribution!

**Pull request merged**